### PR TITLE
node: fix npm -g outdated/update npm (issue #15804)

### LINKED ISSUE #15804: Node / npm package broken
# Please always follow these steps:
- [X] Confirmed this is a problem with `brew install`ing one, specific formula (not cask) and not every time you run `brew`? If it's a general `brew` problem please file this issue at https://github.com/Homebrew/brew/issues/new. If it's a `brew cask` problem please file this issue at https://github.com/Homebrew/caskroom/homebrew-cask/new
- [X] Ran `brew update` and retried your prior step?
- [X] Ran `brew doctor`, fixed all issues and retried your prior step?
- [X] Ran `brew gist-logs <formula>` (where `<formula>` is the name of the formula that failed) and included the output link?
```
brew gist-logs node
Error: No logs.
```
- [X] If `brew gist-logs` didn't work: ran `brew config` and `brew doctor` and included their output with your issue?
```
brew config
HOMEBREW_VERSION: 1.2.4-89-g997260d
ORIGIN: https://github.com/Homebrew/brew.git
HEAD: 997260da541cf4d1be8769e8c7960372bd6a92f0
Last commit: 27 minutes ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 1bf6366af6ded2fbb6fb7627ec1b844848293c08
Core tap last commit: 3 hours ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_REPOSITORY: /usr/local/Homebrew
HOMEBREW_CELLAR: /usr/local/Cellar
HOMEBREW_BOTTLE_DOMAIN: https://homebrew.bintray.com
CPU: octa-core 64-bit ivybridge
Homebrew Ruby: 2.0.0-p648
Clang: 8.1 build 802
Git: 2.13.3 => /usr/local/bin/git
Perl: /usr/bin/perl
Python: /usr/bin/python
Ruby: /Users/John/.rbenv/shims/ruby => /Users/John/.rbenv/versions/2.4.0/bin/ruby
Java: N/A
macOS: 10.12.6-x86_64
Xcode: 8.3.3
CLT: 8.3.2.0.1.1492020469
X11: N/A
```

**Please note we will close your issue without comment if you delete or do not fill out the issue checklist and provide ALL the requested information.**

To help us debug your issue please explain:
- What you were trying to do (and why)
- What happened (include command output)
- What you expected to happen
- Step-by-step reproduction instructions (by running `brew install` commands)

After doing a `brew update` and `brew upgrade`, the latest version of node was installed - v8.2.0. With it came npm V5.3.0.

After this update, npm commands are failing. Doing a `npm -g outdated` gives:

```
~ npm -g outdated
Unhandled rejection Error: ENOENT: no such file or directory, stat '/Users/brew/Library/Caches/Homebrew/node--npm-5.3.0.tgz'

npm ERR! cb() never called!
```

Note that it is looking in `/Users/brew/Library/Caches` .. for files.. obviously there's no such user

### PR DESCRIPTION
**Pull request by chrmoritz**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

This PR fixes #15804 and adds tests for the newly added `npx`.

### CODE REVIEW

**Review by ilovezfs — APPROVED** [review]

File: Formula/node.rb [hunk 1]
```diff
@@ -59,6 +59,11 @@ def install
       bootstrap.install resource("npm")
       system "node", bootstrap/"bin/npm-cli.js", "install", "-ddd", "--global",
              "--prefix=#{libexec}", resource("npm").cached_download
+
+      # The `package.json` stores integrity informations about the above passed
```
Comment (ilovezfs) [inline Formula/node.rb]: information not informations

File: Formula/node.rb [hunk 2]
```diff
@@ -130,8 +135,12 @@ def caveats
...
       assert (HOMEBREW_PREFIX/"bin/npm").exist?, "npm must exist"
       assert (HOMEBREW_PREFIX/"bin/npm").executable?, "npm must be executable"
-      system "#{HOMEBREW_PREFIX}/bin/npm", "--verbose", "install", "npm@latest"
-      system "#{HOMEBREW_PREFIX}/bin/npm", "--verbose", "install", "bignum" unless head?
+      npm_args = ["-ddd", "--cache=#{HOMEBREW_CACHE}/npm_cache", "--build-from-source"]
+      system "#{HOMEBREW_PREFIX}/bin/npm", *npm_args, "install", "npm@latest"
+      system "#{HOMEBREW_PREFIX}/bin/npm", *npm_args, "install", "bignum" unless head?
+      assert (HOMEBREW_PREFIX/"bin/npx").exist?, "npx must exist"
+      assert (HOMEBREW_PREFIX/"bin/npx").executable?, "npx must be executable"
+      assert_match /< hello >/, shell_output("#{HOMEBREW_PREFIX}/bin/npx --cache=#{HOMEBREW_CACHE}/npm_cache cowsay hello")
```
Comment (ilovezfs) [inline Formula/node.rb]: This can be "< hello >"

**Comment by ilovezfs** [comment]
@chrmoritz also we need to bump the revision.

**Comment by chrmoritz** [comment]
fixed your comments

**Comment by chrmoritz** [comment]
fixed comments and bumped formula revision

**Comment by ilovezfs** [comment]
lgtm

**Comment by chrmoritz** [comment]
BTW: I've tried to write a test for `npm -g oudated npm` too, but unfortunately it only exits with status 0 if there is no new version available, but it exits with status 1 in both cases: new version available or any runtime error like ours.

**Comment by ilovezfs** [comment]
@chrmoritz you can use assert_match "foo", shell_output("...", my_expected_exit_code)

**Comment by chrmoritz** [comment]
But we can't predict if a new version will be available at test time (exit code 1) or not (exit code 0) and we can't distinguish between just an update available or a real runtime error from the exit code (in both cases 1). And trying to parse the returned shell_output would be quite complicated, because we would have to support both cases (update available and no update available).

**Comment by ilovezfs** [comment]
Could do blah = Utils.popen_read(...). If blah.include? outdated, assert_match outdated, shell_output(..., 1) else assert_match notoutdated, shell_output(...)

**Comment by chrmoritz** [comment]
Unfortunately the output of `npm outdated` is a table, which is quite hard to predict and each table header could potential be also part of an error message. It wouldn't be easy to match against something to distinguish between an available update and a real npm error.

**Comment by ilovezfs** [comment]
ok

**Comment by ilovezfs** [comment]
Thanks @chrmoritz! :face_with_head_bandage:

**Comment by DomT4** [comment]
Welp. Not having to fiddle with `package.json` didn't last long 😭.

**Pull request closed without merging**